pandas/core/groupby/groupby.py

```diff
@@ -1024,6 +1026,9 @@ def _set_group_selection(self) -> None:
             # GH12839 clear selected obj cache when group selection changes
             ax = self.obj._info_axis
             self._group_selection = ax.difference(Index(groupers), sort=False).tolist()
+            self._group_selection = [
+                idx for idx, label in enumerate(ax) if label not in groupers
+            ]
```

Comment:
I timed this using the code below, this is faster on DataFrames with < 4000 columns, but becomes significantly slower on larger. Going to also post ASVs. Not sure if there is a more efficient method.

Top line is main, bottom line is this PR

```
10
71.8 µs ± 120 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
1.18 µs ± 2.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

100
108 µs ± 428 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
6.47 µs ± 19.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

1000
113 µs ± 2.24 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
63.1 µs ± 238 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

2000
112 µs ± 141 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
129 µs ± 421 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

4000
115 µs ± 129 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
260 µs ± 907 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

8000
127 µs ± 137 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
511 µs ± 1.99 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

10000
132 µs ± 2.66 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
648 µs ± 2.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

100000
349 µs ± 265 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
7.2 ms ± 16.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Code:

```
for size in (10, 100, 1000, 2000, 4000, 8000, 10000, 100000):
    ax = pd.Index(np.random.randint(0, 100, size=size))
    groupers = frozenset(np.random.randint(0, 100, size=5))
    print(size)
    %timeit ax.difference(Index(groupers), sort=False).tolist()
    %timeit [idx for idx, label in enumerate(ax) if label not in groupers]
    print()
```

Reply:
This should be a bit faster: 

```
    y = ax.get_indexer_for(groupers)
    z = np.delete(np.arange(len(ax)), y)
```

Reply:
Another thing: Can't we just drop the duplicates before selecting the columns?

Reply:
Thanks! Still slower than the original, but a definite improvement. Also, the original is doing a different computation, so not good to compare. I would guess typical usage is with a DataFrame with < 100 columns, so it seems beneficial to have the list-comprehension version as well.

> Another thing: Can't we just drop the duplicates before selecting the columns?

We want to keep the duplicates. If the columns (ignoring the grouping) are `['a', 'b', 'a']`, we want to persist these in `DataFrameGroupBy._selected_obj` and in that order. This is why I'm switching from labels to ilocs.

These timings are on my laptop, so a bit slower than the ones posted previously. The third row is @phofl's alternative.

```
10
215 µs ± 9.12 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
2.74 µs ± 16.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
299 µs ± 175 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

100
228 µs ± 5.65 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
15.2 µs ± 51 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
118 µs ± 1.12 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

1000
226 µs ± 2.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
155 µs ± 630 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
175 µs ± 1.42 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

2000
241 µs ± 3.22 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
291 µs ± 1.52 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
246 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

4000
239 µs ± 498 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
655 µs ± 1.65 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
395 µs ± 1.74 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

8000
276 µs ± 5.05 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
1.21 ms ± 1.18 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
652 µs ± 1.55 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

10000
264 µs ± 2.33 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
1.64 ms ± 2.78 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
822 µs ± 8.11 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

100000
619 µs ± 2.72 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
18.2 ms ± 1.77 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.6 ms ± 1.36 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Code:

```
for size in (10, 100, 1000, 2000, 4000, 8000, 10000, 100000):
    ax = pd.Index(np.random.randint(0, 100, size=size))
    groupers = frozenset(np.random.randint(0, 100, size=20))
    print(size)
    %timeit ax.difference(Index(groupers), sort=False).tolist()
    %timeit np.array(idx for idx, label in enumerate(ax) if label not in groupers)
    %timeit np.delete(np.arange(len(ax)), ax.get_indexer_for(list(groupers)))
    print()
```

Reply:
Sorry I meant removing duplicates in _group_selection. They will persist when you do self.obj[self._group_selection] but no further duplication when _group_selection is unique. The order is changed though

Reply:
How did you get the list comprehension so fast? Looks a bit odd that it’s constant

Reply:
I do think we need order to come out the same as the input.

The timings for the list comprehension are the time it takes for the first iteration to fail 🤦 will fix and repost.

Reply:
@phofl - timings have been updated.

Reply:
New implementation: `np.arange(len(ax))[~ax.isin(groupers)]`. Still not as fast as Python lists comprehension for small sizes, but faster than previous. This is faster than main (and now correct), and I think this is quick enough for this portion of the code. For example, for a 10 column 1000 row frame for floats I'm seeing `df.groupby('a').sum()` take 300 microseconds, 20 of which we're spending for this computation.

<details>
<summary>Timings</summary>

```
10
1.84 µs ± 45.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
48.7 µs ± 1.29 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
19.7 µs ± 844 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

100
10.2 µs ± 24.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
50.8 µs ± 121 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
19.6 µs ± 82.6 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

1000
102 µs ± 2.37 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
85.1 µs ± 1.82 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
23.4 µs ± 105 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

2000
200 µs ± 811 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
115 µs ± 1.2 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
27.2 µs ± 110 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

4000
395 µs ± 675 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
190 µs ± 6.03 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
37 µs ± 454 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

8000
808 µs ± 963 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
329 µs ± 202 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
52 µs ± 195 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

10000
1.02 ms ± 22.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
399 µs ± 932 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
69.5 µs ± 4.53 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

100000
10.8 ms ± 621 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
3.58 ms ± 68.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
1.1 ms ± 17.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

</details>

<details>
<summary>Code</summary>

```
for size in (10, 100, 1000, 2000, 4000, 8000, 10000, 100000):
    ax = pd.Index(np.random.randint(0, 100, size=size))
    groupers = frozenset(np.random.randint(0, 100, size=5))
    print(size)
    %timeit np.array([idx for idx, label in enumerate(ax) if label not in groupers])
    %timeit np.delete(np.arange(len(ax)), ax.get_indexer_for(list(groupers)))
    %timeit np.arange(len(ax))[~ax.isin(groupers)]
    print()
```

</details>